stl/inc/atomic

```diff
@@ -639,18 +632,20 @@ struct _Atomic_storage<_Ty, 8> { // lock-free using 8-byte intrinsics
         // non-atomically initialize this atomic
     }
 
-#ifdef _M_IX86
-    void store(const _Ty _Value, const memory_order _Order = memory_order_seq_cst) noexcept {
-        // store with (effectively) sequential consistency
-        _Check_store_memory_order(_Order);
-        (void) exchange(_Value, _Order);
-    }
-#else // ^^^ _M_IX86 / !_M_IX86 vvv
 
     void store(const _Ty _Value) noexcept { // store with sequential consistency
         const auto _Mem           = _Atomic_address_as<long long>(_Storage);
         const long long _As_bytes = _Atomic_reinterpret_as<long long>(_Value);
-#ifdef _M_ARM64
+#if defined(_M_IX86)
+        const auto _Mem32 = _Atomic_address_as<long>(_Storage);
+        // False interlocked modification to provide seq cst fence
+        // The same variable is used as it is known to be cached anyway
+        _InterlockedXor(_Mem32, 0L);
+        // The actual store
+        __iso_volatile_store64(_Mem, _As_bytes);
+        // Another seq cast fence after actual store
+        _InterlockedXor(_Mem32, 0L);

```

Comment:
This looks bad.  This would compile to something with two locked instructions, like this, right?

```
   lock xor  dword ptr [rsp + dummy_local], 0      ; totally pointless
   mov  [actual_address], value
   lock xor  dword ptr [rsp + dummy_local], 0      ; drain the store buffer to promote this from release to seq_cst
```

You only need a compiler barrier before the store; in asm a store can't reorder at runtime with previous loads or stores.  You only need `mov` + memory barrier, *or* make the barrier part of the store by doing it with `xchg`.

I think `xchg` is a good choice; clang does that for seq_cst stores.

Reply:
It is 64-bit store for x86.
Previously it has been done by cmpxchg8b. This whole optimization is about using sse2/floating store. But for seq cst, additional fence is needed over sse2 store

Reply:
In theory I think there could be a mode that produces 32 bit binary and still uses 64 bit instructions, but in practice I think the goal of still supporting x86 is backward compatibility, so no there's no 64-bit exchange here

Reply:
> It is 64-bit store for x86.
> Previously it has been done by cmpxchg8b. This whole optimization is about using sse2/floating store. But for seq cst, additional fence is needed over sse2 store

Oh, I see.  So you can't use xchg, but the other option still applies: store then barrier is all you need.  *Any* store to normal WB memory other than `movnt*` has the same ordering properties as an integer `mov`, including SSE2 `movq` or SSE1 `movlps`.  You want code-gen like `gcc -m32` does;

```
  ; prevent compile-time reordering
   movq  [mem], xmm0
   mfence  or equivalent using lock
```
For correctness it would be fine to do a no-op locked instruction like `lock xor` on the shared variable we just stored, but in practice that's probably a bad idea.  The CPU might share that cache line before lock xor and then have to get it back, making high-contention situations worse.  Better to use a dummy local var.

> a mode that produces 32 bit binary and still uses 64 bit instructions

Only if it switched to a 64-bit code segment like WoW64 DLLs apparently do, and that would only be possible when running under a 64-bit kernel, not in legacy mode.

Unlike when 386 added 32-bit instructions that you could use via prefixes from 16-bit mode, AMD64 had to re-purpose existing 32-bit opcodes as REX prefixes.  So 64-bit operand size (for integer) is only available in the new 64-bit mode.

Reply:
Thank, then I'll `lock inc` local variable after store, which is smaller by one byte than `lock xor`. And put only a compiler barrier before store.

Reply:
I'll actually reuse here `atomic_thread_fence(seq_cst)` which will be `lock inc [esp+constant]` due to another PR

Reply:
Merged with the fix of `atomic_thread_fence`, now calling  `atomic_thread_fence` after `__iso_volatile_store64`. The only thing left I guess is to wait for `__iso_volatile_store64` compiler error fix to arrive (DevCom-986061)